dev/AppNotifications/ShellLocalization.cpp

```diff
@@ -0,0 +1,272 @@
+
+    THROW_IF_FAILED(spStreamOut->SetSize(statstg.cbSize));
+
+    THROW_IF_FAILED(spStream->CopyTo(spStreamOut.get(), statstg.cbSize, nullptr /* pointer to number of bytes read */, nullptr /* pointer to number of bytes written */));
+
+    THROW_IF_FAILED(spStreamOut->Commit(STGC_DANGEROUSLYCOMMITMERELYTODISKCACHE));
+}
+
+HRESULT IsIconFileExtensionSupported(std::filesystem::path const& iconFilePath)
+{
+    const auto fileExtension = iconFilePath.extension();
+
+    std::string lowercaseFileExtension{ fileExtension.u8string() };
+
+    std::transform(lowercaseFileExtension.begin(), lowercaseFileExtension.end(), lowercaseFileExtension.begin(),
+        [](unsigned char c) { return static_cast<char>(std::tolower(c)); });
+
+    const bool isFileExtensionSupported =
+        lowercaseFileExtension == ".ico" || lowercaseFileExtension == ".bmp" || lowercaseFileExtension == ".jpg" || lowercaseFileExtension == ".png";
+
+    THROW_HR_IF(E_UNEXPECTED, isFileExtensionSupported);
+    return S_OK;
+}
+
+inline std::wstring RetrieveLocalFolderPath()
+{
+    wil::unique_cotaskmem_string localAppDataPath;
+    THROW_IF_FAILED(SHGetKnownFolderPath(FOLDERID_LocalAppData, 0 /* flags */, nullptr /* access token handle */, &localAppDataPath));
+
+    // path: C:\Users\<currentUser>\AppData\Local\Microsoft\WindowsAppSDK
+    std::path localFolderPath{ std::wstring(localAppDataPath.get()) +
+        Microsoft::Windows::AppNotifications::Helpers::c_localMicrosoftFolder +
+        Microsoft::Windows::AppNotifications::Helpers::c_localWindowsAppSDKFolder };
+
+    if (!std::exists(localFolderPath))
+    {
+        std::create_directory(localFolderPath);
+    }
+
+    return std::wstring(localFolderPath.c_str());
+}
+
+inline wil::unique_hicon RetrieveIconFromProcess()
+{
+    std::wstring processPath{};
+    THROW_IF_FAILED(wil::GetModuleFileNameExW(GetCurrentProcess(), nullptr, processPath));
+
+    // Extract the small icon from the first .ico, if failed extract the large icon.
+    // Small icon is good enough for an App Notification icon since higher quality doesn't really impact in a substantial way.
+    wil::unique_hicon hIcon{};
+    if (!ExtractIconExW(processPath.c_str(), 0 /* index */, nullptr /* Large icon */, &hIcon, 1))
+    {
+        THROW_HR_IF(E_FAIL, ExtractIconExW(processPath.c_str(), 0, &hIcon, nullptr /* Small Icon */, 1) == 0);
+    }
+
+    return hIcon;
+}
+
+HRESULT Microsoft::Windows::AppNotifications::ShellLocalization::RetrieveAssetsFromProcess(_Out_ Microsoft::Windows::AppNotifications::ShellLocalization::AppNotificationAssets& assets) noexcept try
+{
+    wil::unique_hicon hIcon{ RetrieveIconFromProcess() };
+
+    std::wstring notificationAppId{ Microsoft::Windows::AppNotifications::Helpers::RetrieveNotificationAppId() };
+
+    // path: C:\Users\<currentUser>\AppData\Local\Microsoft\WindowsAppSDK\{AppGUID}.png
+    std::wstring iconFilePath{ RetrieveLocalFolderPath().c_str() + c_backSlash + notificationAppId + c_pngExtension };
```

Comment:
lots of string concatenation makes me sad
Each + is a new copy so lots of malloc+free+memcpy
std::format is your friend...if you've got C++20. We're currently C++17 so wil::str_printf is your friend

BUT...you've building a path. `std::filesystem::path` is actually your friend here

```
auto iconFilePath{ std::filesystem::path(RetrieveLocalFolderPath()) / (notificationAppId + c_pngExtension) };
```

Though it'd be even better if `RetrieveLocalFolderPath` returned `std::filesystem::path`. After all it's called RetrieveLocalFolder**Path** not RetrieveLocalFolder**String** :-)